Add repo_sec_checker tool

jelly opened the pull request:
repo_sec_checker scans all repository's for missing hardening bits. The
tool currently supports relro, stack canary, nx, pie, rpath, runpath and
fortify. The output is currently limited to a table view , which can
later be expanded to JSON output.

jelly commented:
Example output:
```
cdrdao-1.2.4-1-x86_64.pkg.tar.xz                                 Enabled       Enabled   Enabled   Disabled  False    False      Partial: getcwd,strncpy,memset,memcpy,strcpy,read,fread,memmove,strcat
dfu-programmer-0.7.2-2-x86_64.pkg.tar.xz                         Enabled       Enabled   Enabled   Disabled  False    False      Partial: memset,fgets
libgfshare-2.0.0-2-x86_64.pkg.tar.xz                             Enabled       Disabled  Enabled   Disabled  False    False      Partial: memcpy
editorconfig-core-c-0.12.3-1-x86_64.pkg.tar.xz                   Enabled       Enabled   Enabled   Disabled  False    False      Partial: strncpy,strcpy,strncat,fgets,memcpy,memmove,strcat
squashfuse-0.1.103-2-x86_64.pkg.tar.xz                           Enabled       Enabled   Enabled   Disabled  False    False      Partial: memset,memcpy,pread
libsmbios-2.4.2-1-x86_64.pkg.tar.xz                              Enabled       Enabled   Enabled   Disabled  False    False      Partial: strncpy,strcpy,fread,memset,memcpy
libusbmuxd-1.0.10+29+g9db5747-1-x86_64.pkg.tar.xz                Enabled       Enabled   Enabled   Disabled  False    False      Partial: recv,strncpy,memset,read,stpncpy,memmove
```

TODO:

- [ ] check if some performance can be improved
- [ ] add a verbose flag? And hide the missing fortify functions by default
- [ ] make the intermediate format (class) easier to generic json data 
- [ ] add a flag to filter on repository such as core, multilib

SantiagoTorres commented:
I love this!

Foxboron commented:
Should we standardize on underscore or dashes?

jelly commented:
Of yeha let's make it -

Foxboron commented:
- [ ] Fix filename `s/_/-`
- [ ] Probably wants an Author field
- [ ]  Missing SPDX license header

jelly commented:
I've fixed @Foxboron's comments, added RUNPATH/RPATH but they are a bit big. For haskell:
```
In [51]: getattr(tag, 'runpath')                                                                                                                                         
Out[51]: '/usr/lib/ghc-8.6.5/array-0.5.3.0:/usr/lib/ghc-8.6.5/base-4.12.0.0:/usr/lib/ghc-8.6.5/binary-0.8.6.0:/usr/lib/ghc-8.6.5/bytestring-0.10.8.2:/usr/lib/ghc-8.6.5/containers-0.6.0.1:/usr/lib/ghc-8.6.5/deepseq-1.4.4.0:/usr/lib/ghc-8.6.5/directory-1.3.3.0:/usr/lib/ghc-8.6.5/filepath-1.4.2.1:/usr/lib/ghc-8.6.5/ghc-boot-th-8.6.5:/usr/lib/ghc-8.6.5/ghc-prim-0.5.3:/usr/lib/ghc-8.6.5/integer-gmp-1.0.2.0:/usr/lib/ghc-8.6.5/mtl-2.2.2:/usr/lib/ghc-8.6.5/parsec-3.1.13.0:/usr/lib/ghc-8.6.5/pretty-1.1.3.6:/usr/lib/ghc-8.6.5/process-1.6.5.0:/usr/lib/ghc-8.6.5/rts:/usr/lib/ghc-8.6.5/template-haskell-2.14.0.0:/usr/lib/ghc-8.6.5/text-1.2.3.1:/usr/lib/ghc-8.6.5/time-1.8.0.2:/usr/lib/ghc-8.6.5/transformers-0.5.6.2:/usr/lib/ghc-8.6.5/unix-2.7.2.2:/usr/lib'
```

svenstaro commented:
Seems like the nits are pretty minor. Can we just merge it now and then maybe someone else can tidy it up a little if still desired?

jelly commented on security/repo-sec-checker:
```diff
@@ -0,0 +1,247 @@
+def read_file(verbose, filename):
+    res = set()
+
+    with file_reader(filename) as pkg:
```
libarchive has no docs so I have to provide dir() output:
```
['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '_archive_p', '_entry_p', '_getpathname', '_linkpath', '_setpathname', 'atime', 'birthtime', 'ctime', 'filetype', 'get_blocks', 'gid', 'isblk', 'ischr', 'isdev', 'isdir', 'isfifo', 'isfile', 'islnk', 'isreg', 'issock', 'issym', 'linkname', 'linkpath', 'mode', 'mtime', 'name', 'path', 'pathname', 'rdevmajor', 'rdevminor', 'set_atime', 'set_birthtime', 'set_ctime', 'set_mtime', 'size', 'strmode', 'uid']
```

jelly replied:
So it is possible with:
```
+             if entry.mode & S_IXUSR != S_IXUSR:
+                 continue
```

anthraxx commented on security/repo-sec-checker:
```diff
@@ -0,0 +1,250 @@
+from tabulate import tabulate
+
+ARCHS = ['x86_64']
+PKG_EXT = '.tar.xz'
```
seems to miss .zst packages

jelly replied:
ooo good catch, the script is old!

jelly replied:
And updated :)

jelly commented on security/repo-sec-checker:
```diff
@@ -0,0 +1,247 @@
+from multiprocessing import Pool, cpu_count
+from os.path import basename, join
+from re import match, search
+from sys import exit, stderr
```
Fixed

svenstaro commented:
Is this good to go now?

SantiagoTorres commented:
it LGTM now!

Foxboron commented:
Thanks<3
